explorer/interpreter/interpreter.cpp

```diff
@@ -1987,64 +1990,78 @@ auto Interpreter::StepDeclaration() -> ErrorOr<Success> {
+        const auto& member = class_decl.members()[index];
+        if (const auto* var = dyn_cast<VariableDeclaration>(member)) {
+          const Address object = destroy_act.lvalue()->address();
+          const Address var_addr = object.SubobjectAddress(Member(var));
+          const auto v = heap_.Read(var_addr, SourceLocation("destructor", 1));
+          CARBON_CHECK(v.ok())
+              << "Failed to read member `" << var->binding().name()
+              << "` from class `" << class_decl.name() << "`";
+          return todo_.Spawn(std::make_unique<DestroyAction>(
+              arena_->New<LValue>(var_addr), *v));
+        } else {
+          return todo_.RunAgain();
         }
-        // Type of tuple element is integral type e.g. i32
-        // or the type has no destructor
+      } else if (act.pos() == member_count + 1) {
+        if (auto base = class_obj->base()) {
+          return todo_.Spawn(std::make_unique<DestroyAction>(
+              destroy_act.lvalue(), base.value()));
+        } else {
+          return todo_.RunAgain();
+        }
+      } else {
+        todo_.Pop();
+        return Success();
       }
     }
-  }
-
-  if (act.pos() > 0) {
-    if (const auto* class_obj =
-            dyn_cast<NominalClassValue>(destroy_act.value())) {
-      const auto& class_type = cast<NominalClassType>(class_obj->type());
-      const auto& class_dec = class_type.declaration();
-      int index = class_dec.members().size() - act.pos();
-      if (index >= 0 && index < static_cast<int>(class_dec.members().size())) {
-        const auto& member = class_dec.members()[index];
-        if (const auto* var = dyn_cast<VariableDeclaration>(member)) {
-          Address object = destroy_act.lvalue()->address();
-          Address mem = object.SubobjectAddress(Member(var));
-          SourceLocation source_loc("destructor", 1);
-          auto v = heap_.Read(mem, source_loc);
-          return todo_.Spawn(
-              std::make_unique<DestroyAction>(destroy_act.lvalue(), *v));
+    case Value::Kind::TupleValue: {
+      const auto* tuple = cast<TupleValue>(destroy_act.value());
+      const auto element_count = tuple->elements().size();
+      if (static_cast<size_t>(act.pos()) < element_count) {
+        const size_t index = element_count - act.pos() - 1;
+        const auto& item = tuple->elements()[index];
+        const auto object_addr = destroy_act.lvalue()->address();
+        auto* tuple_field =
+            arena_->New<IndexedValue>(IndexedValue{index, item});
+        Address field_address =
+            object_addr.SubobjectAddress(Member(tuple_field));
+        if (item->kind() == Value::Kind::NominalClassValue ||
+            item->kind() == Value::Kind::TupleValue) {
+          return todo_.Spawn(std::make_unique<DestroyAction>(
+              arena_->New<LValue>(field_address), item));
+        } else {
+          // Type of tuple element is integral type e.g. i32
+          // or the type has no destructor
```

Comment:
```suggestion
          // The tuple element's type is an integral type (e.g., i32)
          // or the type doesn't support destruction.
```

Trying to fix punctuation but also clarify a little: destructors are optional on classes, but a class without a destructor still needs DestroyAction run on it.

This does make me wonder if some kind of is_destructible() attribute might add some helpful guard rails, but that's just me musing, probably not worth doing right now.

Reply:
Yeah it might be... and to be fair I kept this comment as-is from existing code, and haven't delved into what was appropriate or not in that regard. On the surface I would tend to agree with your comment.
Done for the comment changes.